### PR TITLE
Excise GroupsCore dependency

### PR DESCRIPTION
**Pull request by fingolfin**
This is work in progress. Among other things, we need to add a few
of generic fallbacks that were provided previously by GroupsCore.

Closes #1525

### CODE REVIEW

File: test/Groups-conformance-tests.jl
Comment (fingolfin) [inline test/Groups-conformance-tests.jl:3]: I disabled these iteration tests for now as we don't implement all of that generically right now, and I am not sure we really want it that way.

File: test/Groups-conformance-tests.jl
Comment (fingolfin) [inline test/Groups-conformance-tests.jl:60]: This test case seems problematic; what if `G` has order greater than `typemax(Int16)`? Probably should instead first get `order(BigInt, G)` and check if that is small enough to fit into an `Int16` (and if it doesn't fit, then perhaps it should `@test_throws order(Int16, G)` instead?)

File: src/Groups.jl
Comment (lgoettgens) [inline src/Groups.jl:56]:
```suggestion
The default implementation is based on `is_finite` and `order`.
```

File: src/Groups.jl
Comment (lgoettgens) [inline src/Groups.jl:67]: This include should not be here and moved to the main file instead.

File: test/Groups-conformance-tests.jl
Comment (lgoettgens) [inline test/Groups-conformance-tests.jl:104]: two equal tests

File: test/Groups-conformance-tests.jl
Comment (lgoettgens) [inline test/Groups-conformance-tests.jl:124]: generic implementations for most of these are still missing.

File: test/Groups-conformance-tests.jl
Comment (lgoettgens) [inline test/Groups-conformance-tests.jl:170]: same problem as above

**Review by lgoettgens — COMMENTED** [review]
For many functions, it is not clear (to me) what the interface expects from a concrete implementation, and what it then provides using generic functions. I would really like to see this written up in the docs, like https://github.com/Nemocas/AbstractAlgebra.jl/blob/master/docs/src/ring_interface.md for the ring interface.
Speaking of it, I think the group interface should be as similar as possible to the ring interface, so we do not create confusion for users.

**Comment by ThomasBreuer** [comment]
Changing the current dependency to GroupsCore 0.4 is necessary in order to be able to use both AbstractAlgebra.jl (for example as a part of Oscar.jl) and Chevie.jl (formerly called Gapjm.jl), because some dependency of Chevie.jl needs GroupsCore 0.5.
Removing the dependency completely from AbstractAlgebra.jl would of course be fine.

File: src/Aliases.jl
Comment (fingolfin) [reply src/Aliases.jl:86]: Couldn't we remove almost *all* `@deprecate_binding` with that argument, as for most of them we had multiple breaking releases since we deprecated?

But  that is only if one believes that semver magically solves the problem of compatibility. But it doesn't: users still have code using the old names; if their code uses `istrivial`, they'll notice if it is gone, semver or not. On the other hand, I am convinced virtually nobody will care or even notice about GroupsCore not being involved anymore...

File: src/Groups.jl
Comment (fingolfin) [reply src/Groups.jl:10]: This is out of scope for this PR, which is about getting rid of GroupsCore with minimal work. Any renaming or refactoring this enables for us can be done in a later commit, by anyone interested in investing time into it... But for now it'd be better to just get this done

File: src/Groups.jl
Comment (fingolfin) [reply src/Groups.jl:23]: Agreed! `eltype` is for iteration and I deliberately stripped out the iteration stuff from GroupsCore (I am happy to have as many group types as possible allow for iteration, I just don't think making it a central and mandatory aspect for defining groups is beneficial)

File: src/Groups.jl
Comment (fingolfin) [reply src/Groups.jl:10]: Actually I had a closer look and it turns out this is reference to as `GroupsCore.InfiniteOrder` everywhere, so we have to change those places, so I might as well rename it now.

Comment (fingolfin) [reply src/Groups.jl:10]: I wonder if we should export `InfiniteOrderError`...

File: src/Groups.jl
Comment (fingolfin) [reply src/Groups.jl:29]: thanks

File: src/Groups.jl
Comment (fingolfin) [reply src/Groups.jl:40]: fine by me

File: src/Groups.jl
Comment (fingolfin) [reply src/Groups.jl:56]: Fine by me (though for things like `isfinite = is_finite` that come from `Base`, we have been inconsistent with which spelling to use; there are arguments either way).

Seperately, the comment is incomplete, as it also uses `hasgens` and `gens`

File: src/Groups.jl
Comment (fingolfin) [reply src/Groups.jl:67]: Yes but that's a separate issue that affects a bunch of other files. I'll untangle it in a different PR

File: test/Groups-conformance-tests.jl
Comment (fingolfin) [reply test/Groups-conformance-tests.jl:45]: agreed

File: src/Groups.jl
Comment (lgoettgens) [reply src/Groups.jl]: I would say yes. It would be strange, if some Oscar group (e.g. Weyl group of a generalized Cartan matrix) throws an `AbstractAlgebra.InfiniteOrderError` instead of an `InfiniteOrderError`.

File: test/Groups-conformance-tests.jl
Comment (fingolfin) [inline test/Groups-conformance-tests.jl:149]: In GroupsCore, we had `commutator`, which we also imported; but we didn't use this anywhere. This is in line with our agreement to have "self-explanatory" names in general (though we sometimes make exceptions)

In Oscar we have `comm`, which is also used there. This is kinda in line with `inv` and `conj` and `gens` ... I think I am going to use that here for now, and leave the decision on whether to rename some or all of these to another time

File: src/Groups.jl
Comment (fingolfin) [reply src/Groups.jl]: `is_finiteorder` or `is_finite_order` ?

Comment (lgoettgens) [reply src/Groups.jl]: `is_finiteorder` is the one used in OSCAR

Comment (fingolfin) [reply src/Groups.jl]: Sure, but logically, it would be `is_finite_order`... anyway, I left it at the `is_finiteorder` for now to simplify things

**Comment by fingolfin** [comment]
This is ready now. OscarCI tests of course fail because they still reference `GroupsCore`. I'll prepare a (draft) PR for Oscar now which adapt to this patch, to verify it is actually viable.

File: src/AbstractAlgebra.jl
Comment (fingolfin) [inline src/AbstractAlgebra.jl:314]: I previously merged these `export` statements with the list below, but that broke Hecke: because the `@alias` statements ensure that if an alias is exported, also the "original" is exported and vice versa, and apparently Hecke relies on AA exporting (at least) `isone`... We should add an `import Base.isone` to Hecke instead to make it more resilient (but also should move all exports to the *top* export list in this file)

**Comment by lgoettgens** [comment]
> This is ready now. OscarCI tests of course fail because they still reference `GroupsCore`. I'll prepare a (draft) PR for Oscar now which adapt to this patch, to verify it is actually viable.

I'll do a detailed review tomorrow.

File: src/Groups.jl
Comment (lgoettgens) [inline src/Groups.jl:68]: Throwing `function order(::Type{T}, G::GroupElem) where T` is missing. (For consistency with the one for Groups.)

File: src/Groups.jl
Comment (lgoettgens) [inline src/Groups.jl:84]: reference to GroupsCore

File: src/Groups.jl
Comment (lgoettgens) [inline src/Groups.jl:128]: The docstring should include the fact, that this is `true` by default and needs to be changed by concrete implementations (if needed).

**Review by lgoettgens — COMMENTED** [review]
I am missing a mention of `is_finite(::Group)` and `gen(::Group, ::Int)` in `Groups.jl` at the list of throwing function stubs.

File: src/Groups.jl
Comment (fingolfin) [reply src/Groups.jl]: fixed

File: src/Groups.jl
Comment (fingolfin) [reply src/Groups.jl:65]: Added. I also revised the docstring for both `order` methods, please check.

File: src/Groups.jl
Comment (lgoettgens) [inline src/Groups.jl:91]: This should be `NotImplementedError` with different args

File: src/Groups.jl
Comment (lgoettgens) [inline src/Groups.jl:110]:
```suggestion
The default implementation is based on `gens`, `is_finite` and `order`.
```

File: src/Groups.jl
Comment (lgoettgens) [inline src/Groups.jl:114]:
```suggestion
    is_finite(G) && return isone(order(G))
```

File: src/Groups.jl
Comment (lgoettgens) [reply src/Groups.jl:137]: I still think that this either shouldn't be true by default or at least stated so in the docstring

Comment (lgoettgens) [reply src/Groups.jl:137]: I still think that this either shouldn't be true by default or at least stated so in the docstring

**Review by lgoettgens — COMMENTED** [review]
Some more comments.

I still would like to have a docpage where the expectations and features of the interface are listed, similar to e.g. the Ring interface. This doesn't need to be as lengthy as the already existing ones, but I think even just two short lists (needed by a concrete implementation, generic interface functionality) in a md file that is linked in the docs is better than having just nothing.

File: src/Groups.jl
Comment (fingolfin) [reply src/Groups.jl]: done

File: src/Groups.jl
Comment (fingolfin) [reply src/Groups.jl]: Rewrote this part of the docstring

File: src/Groups.jl
Comment (fingolfin) [reply src/Groups.jl]: done

File: src/Groups.jl
Comment (fingolfin) [reply src/Groups.jl:138]: I've added a note stating what the default is.

**Comment by fingolfin** [comment]
Rebased and addressed all open comments (or at least I tried).

**Review by lgoettgens — APPROVED** [review]
I have no further objections to this. Some things in the conformance tests, I maybe want to change in the future, but that should be non-breaking.

And a big wish is a docpage about this, but this shouldn't block this PR. Best I should open a issue once this is merged.

**Pull request merged**
